Offline: Plumb clientSequenceNumber from loader submit APIs to PendingStateManager again

markfields opened the pull request:
## Description

Fixes [AB#8282](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8282).  It's a partial revert of #13994.

The Container's various "submit" functions all return a client sequence number (CSN) which is assigned sequentially when the loader layer sends the message to the ordering service.  For a batch, the CSN returned is the CSN of the last message in the batch, including chunks sent during chunking (if applicable).

We take this and convert it to the CSN of the start of the back, which will be used to form a new Batch ID for detecting duplicate changes across container forks, e.g. when improperly rehydrating from a serialized container.

markfields committed:
Cherry-pick CSN changes from prototype branch (c9ba7be302)

markfields committed:
Send messages to PSM one batch at a time

This lets us be simpler/explicit about only having the CSN for the start of the batch (not reconstructing CSNs for each subsequent message)

markfields committed:
api.md miss from cherry-pick

markfields committed:
tweaks

markfields committed:
Deal with //*

markfields committed:
Tests, nits, clean up

markfields committed:
Merge branch 'main' into o3/csn

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/opLifecycle/outbox.ts:
```diff
@@ -380,10 +387,10 @@ export class Outbox {
 			);
 
-			this.params.legacySendBatchFn(batch);
+			clientSequenceNumber = this.params.legacySendBatchFn(batch);
```
Do we know for sure this returns clientSequenceNumber?

markfields replied:
Yes - it was like this before #13994 

And for fun I browsed back via git blame... 2019 was the last time I can't immediately prove it returns a CSN.
https://github.com/microsoft/FluidFramework/pull/839
https://github.com/microsoft/FluidFramework/commit/05a5df0106238896ac2913d2bae356ccd3cde4e3

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/opLifecycle/outbox.ts:
```diff
@@ -380,10 +387,10 @@ export class Outbox {
 		} else {
 			assert(batch.referenceSequenceNumber !== undefined, 0x58e /* Batch must not be empty */);
-			this.params.submitBatchFn(
+			clientSequenceNumber = this.params.submitBatchFn(
```
Same here

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/pendingStateManager.ts:
```diff
@@ -270,6 +275,18 @@ export class PendingStateManager implements IDisposable {
+		// The server sequences messages in the order received, just like the pending message queue
+		if (this.pendingBatchBeginMessage?.clientSequenceNumber !== pendingMessage.batchStartCsn) {
+			this.logger?.sendErrorEvent({
+				eventName: "PendingClientSequenceNumberMismatch",
```
Looks like some test is hitting this error

markfields replied:
Yeah had 2 bugs. One around grouped batching and the other around single-message batches of any kind.

I'm going to have to remove this for now, the fix for the grouped batching one deserves its own PR (and adding this check is not a priority - it was a learning opportunity for me)

markfields replied:
FYI @dannimad

markfields committed:
Fix 2 CSN validation bugs

1. Single-message batches weren't getting the incoming CSN properly
2. Grouped batching re-uses CSN, losing the batch's original CSN

The second one is fixed by a semi-hack (on top
of the existing CSN hack), so going to revert that next.

markfields committed:
Revert PendingClientSequenceNumberMismatch

It was complicated / compromised by Grouped batching's use of fake CSN.
I think it's doable (see change to opGroupingManager here)
but want to separate it from current PR.

markfields committed:
nit

markfields committed:
Build fixes

markfields committed:
Fix outbox UTs

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/pendingStateManager.ts:
```diff
@@ -187,27 +189,30 @@ export class PendingStateManager implements IDisposable {
+	public onFlushBatch(batch: BatchMessage[], clientSequenceNumber: number | undefined) {
+		for (const message of batch) {
+			const {
+				contents: content = "",
```
Why do we need this "" here? This seems odd when considering we didn't have it before in the previous `onSubmitMessage`

markfields replied:
See old outbox line 404, it was doing ! 

I have a change stashed that fixes the types here to align.

tyler-cai-microsoft replied:
Maybe make this an assert, not sure

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/pendingStateManager.ts:
```diff
@@ -187,27 +189,30 @@ export class PendingStateManager implements IDisposable {
+	 * @param clientSequenceNumber - The CSN of the first message in the batch,
+	 * or undefined if the batch was not yet sent (e.g. by the time we flushed we lost the connection)
 	 */
-	public onSubmitMessage(
```
Maybe this is me overthinking things, (non-blocking) why don't we just keep `onSubmitMessage` and call `onSubmitMessage` during `onFlushBatch` or even convert `onSubmitMessage` to `private`?

markfields replied:
Direction is to have PSM deal in while batches not individual messages

tyler-cai-microsoft reviewed (approved):
LGTM

dannimad reviewed (approved):
LGTM as well. Just curious about why you moved some batch logic to the pending state manager. It looks to me that batch logic could have been left on outbox and psm keep talking in ops. Is it some kind of intermediate step towards changing the whole pending ops processing?
